src/process/script.h

```diff
@@ -0,0 +1,257 @@
+  if (buffer == NULL) return;
+
+  interpreter_path_t *path = (interpreter_path_t *)buffer;
+
+  // skip check if sz == BINPRM_BUF_SIZE. If a future kernel increases
+  // the value of BINPRM_BUF_SIZE then we may run into truncation but
+  // I don't think we need to worry right now.
+  ret = bpf_probe_read_str(&path->path, BINPRM_BUF_SIZE, interp);
+  if (ret < 0) {
+    set_empty_local_warning(PMW_READ_PATH_STRING);
+    return;
+  }
+
+  if (path->path[0] == '/') {
+    ret = bpf_map_update_elem(&interpreters, new_interpreter, path, BPF_ANY);
+    if (ret < 0) {
+      error_info_t info = {0};
+      info.err = ret;
+      set_local_warning(PMW_UPDATE_MAP_ERROR, info);
+      goto EmitWarning;
+    }
+  } else {
+    relative_file_info_t rel_interpreter = {0};
+    rel_interpreter.pid = pid;
+    rel_interpreter.interpreter = *new_interpreter;
+
+    ret = bpf_map_update_elem(&rel_interpreters, &rel_interpreter, path, BPF_ANY);
+    if (ret < 0) {
+      error_info_t info = {0};
+      info.err = ret;
+      set_local_warning(PMW_UPDATE_MAP_ERROR, info);
+      goto EmitWarning;
+    }
+  }
+
+  return;
+
+ EmitWarning:;
+  process_message_t pm = {0};
+  push_warning(ctx, &pm, PM_SCRIPT);
+
+  return;
+}
+
+static __always_inline u64 push_scripts(struct pt_regs *ctx, buf_t *buffer) {
+  process_message_t *pm = (process_message_t *)buffer;
+  u64 pid_tgid = bpf_get_current_pid_tgid();
+  u32 pid = pid_tgid >> 32;
+  u64 event_id = 0;
+
+  script_t *script = bpf_map_lookup_elem(&scripts, &pid);
+  if (script == NULL) return event_id;
+
+  if (script->interpreters[0].inode == 0) goto Done;
+  // clear out the buffer
+  __builtin_memset(&pm->u.script_info.scripts, 0, sizeof(pm->u.script_info.scripts));
+
+  pm->type = PM_SCRIPT;
+  pm->u.script_info.scripts[0] = script->file.identity;
+  pm->u.script_info.buffer_length = sizeof(process_message_t);
+
+  // the original script file path came from the user; the kernel
+  // verifies that the user provided path cannot go beyond
+  // PATH_MAX. The total path can still be beyond PATH_MAX once you
+  // combine it with the CWD but userspace can take care of that
+  int sz = write_string(script->file.path, buffer, &pm->u.script_info.buffer_length, PATH_MAX);
+  if (sz < 0) goto WriteError;
+
+#pragma unroll MAX_INTERPRETERS
+  for (int i = 1; i < MAX_INTERPRETERS; i++) {
+    // there is no interpreter which means the last interpreter was
+    // the actual executable
+    if (script->interpreters[i].inode == 0) break;
+
+    // The last interpreter was a script
+    file_info_t *intp_key = &script->interpreters[i - 1];
+    pm->u.script_info.scripts[i] = *intp_key;
+
+    relative_file_info_t rel_interpreter = {0};
+    rel_interpreter.pid = pid;
+    rel_interpreter.interpreter = *intp_key;
+
+    interpreter_path_t *intp_path = bpf_map_lookup_elem(&rel_interpreters, &rel_interpreter);
+    if (intp_path == NULL) {
+      intp_path = bpf_map_lookup_elem(&interpreters, intp_key);
+    }
```

Comment:
It seems to me that we're far more likely to find it in the `interpreters` map rather than `rel_interpreters`; does it make sense to check `interpreters` first so we won't have to perform 2 map lookups as often?

Reply:
My only concern with that is that it leaves us potentially open to a very unlikely but possible error case:

1. A process starts a script with an interpreter in path `/my/absolute/path` with file identity X. Because the path is absolute it gets saved inside `/my/absolute/path`. This works as expected.
2. `/my/absolute/path` is renamed to `/my/absolute/interpreter` and the kernel happens to reuse the inode. We don't invalidate the path in the `interpreters` map because we haven't implemented that yet.
3. A process starts a script with cwd `/my/absolute` and the interpreter is `./interpreter`. Because it is relative the path gets saved inside `rel_interpreters` instead of `interpreters`. If we first look for it inside `interpreters` then we'll find the *old* path which is *incorrect*. 

So even though we are more likely to find the path inside `interpreters` rather than `rel_interpreters` if the file identity exists in both for a given PID then the one in `rel_interpreters` takes priority because it is more guaranteed to be newer. 

That being said I just spotted an issue with the current flow because `rel_interpreters` was not being invalidated so I just pushed a change to remove `rel_interpreters` when a new script is run using an absolute path for the same file identity + pid (e.g., double exec). Pretty unlikely to happen but just in case.

Reply:
Based on my experimentation so far with filemod events, renaming via `renameat2` (which is what you get with `mv`) does re-use the inode, and you can also see inode re-use fairly easily with deletion followed by creation of a new item. So this all seems reasonable to me.